File: ios/Classes/FlutterSegmentationDataCallback.swift
//
//  FlutterSegmentationDataCallback.swift
//  exponea
//

import Foundation
import ExponeaSDK

class FlutterSegmentationDataCallback {
    let instanceId = UUID().uuidString

    let exposingCategory: SegmentCategory
    let includeFirstLoad: Bool
    let onNewData: (FlutterSegmentationDataCallback, [SegmentDTO]) -> Void

    lazy var nativeCallback: SegmentCallbackData = {
        return .init(category: exposingCategory, isIncludeFirstLoad: includeFirstLoad) { data in
            Exponea.logger.log(
                .verbose,
                message: "Segments: New segments for '\(self.exposingCategory)' received: \(data)"
            )
            self.onNewData(self, data)
        }
    }()

    init(
        category: SegmentCategory,
        includeFirstLoad: Bool,
        onNewData: @escaping (FlutterSegmentationDataCallback, [SegmentDTO]) -> Void
    ) {
        self.exposingCategory = category
        self.includeFirstLoad = includeFirstLoad
        self.onNewData = onNewData
    }
}

File: android/src/main/kotlin/com/exponea/data/FlutterSegmentationDataCallback.kt
package com.exponea.data

import com.exponea.sdk.models.Segment
import com.exponea.sdk.models.SegmentationDataCallback
import com.exponea.sdk.util.Logger
import java.util.UUID

class FlutterSegmentationDataCallback(
    override val exposingCategory: String,
    override var includeFirstLoad: Boolean,
    private val callback: (FlutterSegmentationDataCallback, List<Segment>) -> Unit
) : SegmentationDataCallback() {
    val instanceId = UUID.randomUUID().toString()

    override fun onNewData(segments: List<Segment>) {
        Logger.d(this, "Segments: New segments for '$exposingCategory' received: $segments")
        callback.invoke(this, segments)
    }
}